prow/crier/controller_test.go

```diff
@@ -0,0 +1,268 @@
+	return f.shouldReportFunc(pj)
+}
+
+// Fake Informer
+// Sets: The Prow Job Test Cases
+type fakeInformer struct {
+	jobs map[string]*prowv1.ProwJob
+}
+
+func (f fakeInformer) Get(name string) (*prowv1.ProwJob, error) {
+	pj, found := f.jobs[name]
+	if !found {
+		var s schema.GroupResource
+		return nil, errors.NewNotFound(s, "Can't Find ProwJob")
+	}
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) prowLister.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() prowLister.ProwJobLister {
+	return f
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) AddEventHandler(handler cache.ResourceEventHandler)                        {}
+func (fakeInformer) Run(stopCh <-chan struct{})                                                {}
+func (fakeInformer) AddEventHandlerWithResyncPeriod(cache.ResourceEventHandler, time.Duration) {}
+func (fakeInformer) GetStore() cache.Store                                                     { return nil }
+func (fakeInformer) GetController() cache.Controller                                           { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                           { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                                 { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                                 { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*prowv1.ProwJob, err error)          { return nil, nil }
+
+func TestController_Run(t *testing.T) {
+	tests := []struct {
+		name         string
+		jobsOnQueue  []string
+		knownJobs    map[string]*prowv1.ProwJob
+		shouldReport func(*prowv1.ProwJob) bool
+
+		expectReports []string
+		expectPatch   bool
+	}{
+		{
+			name:        "reports/patches known job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+			expectReports: []string{"foo"},
+			expectPatch:   true,
+		},
+		{
+			name:        "doesn't report when it shouldn't",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return false
+			},
+		},
+		{
+			name:        "doesn't report nonexistant job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs:   map[string]*prowv1.ProwJob{},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+		//{
+		//	name: "nil job panics",
```

Comment:
why is this commented out?

Reply:
I don't think it's supposed to panic. I think this is a bug.
@krzyzacy , should I write a test for this case?

Reply:
hummm, when will you get a nil prowjob?

Reply:
Maybe never? And in that case, this seems the best way to call out that there's an artificial edge case that causes a panic, this is an easy way to recreate it, and "fixing it" isn't a priority.

Testing and expecting a panic isn't right, because that implies that panicking is correct. "Fixing it" as part of this commit is also bad; not modifying the code is the only way to prove that it will still work.

/shrug